### PR TITLE
Shard -> Gem

### PR DESCRIPTION
**Pull request by jk13pclick**
Building off of Ultima's original PR, this is done. There is a kinda bad solution for the challenge mode, but it's alright.

### CODE REVIEW

File: src/scripts/Update.ts [hunk 1]
```diff
@@ -598,6 +599,47 @@ class Update implements Saveable {
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 32, 34); // New Mauville
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 84, 64); // Pledge Grove
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 78, 79); // Abundant Shrine
+
+            // Shards -> Gems
+            delete saveData.quests.questList;
+
+            //Setting gems = shards
+            saveData.gems = saveData.shards || {};
```
Comment (RedSparr0w) [inline src/scripts/Update.ts:607]: We probably don't need this line, as if `saveData.shards` doesn't exist, the next few lines will fail when it calls `saveData.shards.shardWallet` etc.

File: src/scripts/Update.ts [hunk 2]
```diff
@@ -598,6 +599,47 @@ class Update implements Saveable {
...
+            delete saveData.keyItems['Shard_case'];
+
+            // Swapping Shard Case for Gem Case
+            if (saveData.badgeCase[8]) {
+                saveData.keyItems['Gem_case'] = true;
+            }
+
+            // Just incase statistics is not set
+            saveData.statistics = saveData.statistics || {};
+
```
Comment (RedSparr0w) [inline src/scripts/Update.ts:624]: [Same thing as above](https://github.com/pokeclicker/pokeclicker/pull/1849/files?short_path=10b250f#r821352191)

File: src/scripts/Update.ts [hunk 3]
```diff
@@ -598,6 +599,47 @@ class Update implements Saveable {
...
+            if (saveData.gems.gemUpgrades.every((g: number) => !g)) {
+                Notifier.notify({
+                    title: 'Active Challenge Mode?',
+                    message: `Do you want to activate the No Gem challenge mode?
+                    (Synonymous with the former "No Shard" Challenge)
+
+                    <button class="btn btn-block btn-danger" onclick="App.game.challenges.list.disableGems.activate();" data-dismiss="toast">Activate</button>`,
+                    timeout: GameConstants.HOUR,
+                });
+            }
```
Comment (RedSparr0w) [inline src/scripts/Update.ts:642]: Can't we just check if the shard challenge is enabled/disabled and use the same value?

File: src/scripts/Update.ts [hunk 4]
```diff
@@ -598,6 +599,47 @@ class Update implements Saveable {
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 32, 34); // New Mauville
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 84, 64); // Pledge Grove
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 78, 79); // Abundant Shrine
+
+            // Shards -> Gems
+            delete saveData.quests.questList;
```
[the remark this answers is not in the document]
Comment (jk13pclick) [reply src/scripts/Update.ts:604]: Probably not, if we can just transfer the gainShardQuest data into gainGemQuest data

File: src/scripts/Update.ts (hunk 1 above)
Comment (jk13pclick) [reply src/scripts/Update.ts]: [b191e61](https://github.com/pokeclicker/pokeclicker/pull/1791/commits/b191e611f9850991637fe4ccd96cc1a4bea10577)

File: src/scripts/Update.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jk13pclick) [reply src/scripts/Update.ts]: See [bb44b0c](https://github.com/pokeclicker/pokeclicker/pull/1791/commits/bb44b0c5bd19f492c0e59fd6773e777cdbc0e23c). Unsure how to transfer the completion data

File: src/scripts/Update.ts [hunk 5]
```diff
@@ -598,6 +599,49 @@ class Update implements Saveable {
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 32, 34); // New Mauville
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 84, 64); // Pledge Grove
             saveData.statistics.dungeonsCleared = Update.moveIndex(saveData.statistics.dungeonsCleared, 78, 79); // Abundant Shrine
+
+            // Shards -> Gems
+            saveData.quests.questList = {
+                ...saveData.quests.questList,
+                gainGemsQuest: saveData.quests.questList.gainShardsQuest || 0,
+            };
```
Comment (RedSparr0w) [inline src/scripts/Update.ts:607]: The `questList` is an array of quest, each with a structure like the following:
```json
{
  "index": 0,
  "data": [
    319,
    1250,
    10
  ],
  "initial": 208068,
  "claimed": true,
  "notified": true,
  "name": "GainShardsQuest"
}
```
You would basically just need to loop through it and replace any name instance of `GainShardsQuest` with `GainGemsQuest`

Something like this _should_ work, I haven't tested it though:
```js
saveData.quests.questList = saveData.quests.questList.map(q => {
  if (q.name == 'GainShardsQuest') q.name = 'GainGemsQuest';
  return q;
})
```

File: src/scripts/Update.ts (hunk 3 above)
Comment (RedSparr0w) [reply src/scripts/Update.ts:644]: This _should_ work (again untested):
```js
saveData.challenges.list.disableGems = saveData.challenges?.list?.disableShards || false;
```

File: src/scripts/Update.ts (hunk 5 above)
Comment (jk13pclick) [reply src/scripts/Update.ts]: Works great!

File: src/scripts/Update.ts (hunk 3 above)
Comment (jk13pclick) [reply src/scripts/Update.ts]: Works great!

**Review by RedSparr0w — APPROVED** [review]
Looks good, thanks for the PR!

**Pull request merged**